### PR TITLE
Fix #592: allow multiple readers but no multiple scans on any reader

### LINKED ISSUE #592: Decide whether multiple scanners are allowed
Currently as spec'ed each reader instance will receive the same data.

I am not a bit fan of this. iOS allows the same, but there is an order to it, like the first (or maybe it was the last) added reader receives the data first and if it doesn't handle it, it will be given to the next reader. That makes more sense to be, but this is the web so we really don't want other libraries etc to do any reading, so I think it makes more sense for the call to `scan` to actually cancel existing scans.

@riju @beaufortfrancois

### PR DESCRIPTION
**Pull request by zolkis**
Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/611.html" title="Last updated on Dec 28, 2020, 9:11 AM UTC (341f824)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/611/0a0a609...zolkis:341f824.html" title="Last updated on Dec 28, 2020, 9:11 AM UTC (341f824)">Diff</a>

### CODE REVIEW

**Review by kenchris — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -2225,6 +2225,28 @@ <h2>The <dfn>record type</dfn> string</h2>
...
+  <section><h3>Aborting pending scan operation</h3>
+  <div>
+    To attempt to <dfn>abort a pending scan operation</dfn> on an
+    <a>environment settings object</a>, perform the following steps:
+    <ol class=algorithm>
+      <li>
+        If there is no <a>pending read tuple</a> |tuple|, abort these steps.
+      </li>
+      <li>
+        If |tuple|'s reader has already initiated an ongoing NFC data transfer,
```
Comment (beaufortfrancois) [inline index.html]: Can we link to the definition of an "ongoing NFC data transfer"?

Comment (beaufortfrancois) [inline index.html]: Where is defined |tuple|?

File: index.html [hunk 2]
```diff
@@ -2225,6 +2225,28 @@ <h2>The <dfn>record type</dfn> string</h2>
   </div>
   </section>
 
+  <section><h3>Aborting pending scan operation</h3>
+  <div>
+    To attempt to <dfn>abort a pending scan operation</dfn> on an
+    <a>environment settings object</a>, perform the following steps:
```
Comment (beaufortfrancois) [inline index.html]: Why using "environment settings object" ?

**Review by beaufortfrancois — COMMENTED** [review]
I'm confused with the algorithm. 
See my notes.

File: index.html (hunk 1 above)
Comment (zolkis) [reply index.html]: Two lines above.

File: index.html (hunk 2 above)
Comment (zolkis) [reply index.html]: We do the same for writing. What is the current terminology? We attached the slots to that object.

Comment (zolkis) [reply index.html]: Removed them both. The algorithm should be clear on its own.

File: index.html [hunk 3]
```diff
@@ -2225,6 +2225,28 @@ <h2>The <dfn>record type</dfn> string</h2>
...
+      </li>
+      <li>
+        If |tuple|'s reader has already initiated an ongoing NFC data transfer,
+        abort these steps.
+      </li>
+      <li>
+        Reject |tuple|'s promise with an {{"InvalidStateError"}} {{DOMException}}.
+      </li>
+      <li>
+        Clear the <a>pending read tuple</a> and abort these steps.
```
Comment (beaufortfrancois) [inline index.html]: Why aborting steps at the end?

File: index.html (hunk 1 above)
Comment (beaufortfrancois) [reply index.html]: Got it.

File: index.html [hunk 4]
```diff
@@ -2225,6 +2225,28 @@ <h2>The <dfn>record type</dfn> string</h2>
   </div>
   </section>
 
+  <section><h3>Aborting pending scan operation</h3>
+  <div>
+    To attempt to <dfn>abort a pending scan operation</dfn>,
```
Comment (beaufortfrancois) [inline index.html]: Shall we run "abort a pending scan operation" as well when NFC is suspended?
Same way we do for Write?

Comment (zolkis) [reply index.html]: Good point. We did, but now we can replace 2 steps with one. Coming.

File: index.html (hunk 3 above)
Comment (zolkis) [reply index.html]: Fixed.

File: index.html (hunk 4 above)
Comment (zolkis) [reply index.html]: Also found another error in aborting scan, fixed.

File: index.html (hunk 1 above)
Comment (zolkis) [reply index.html]: We don't do that for writes, either. How can we define it better?

File: index.html [hunk 5]
```diff
@@ -2238,14 +2263,10 @@ <h2>The <dfn>record type</dfn> string</h2>
       Attempt to <a>abort a pending write operation</a>.
     </li>
     <li>
-      Stop the <a>dispatch NFC content</a> steps.
-    </li>
-    <li>
-      Clear the <a>pending read tuple</a>.
+      Attempt to <a>abort a pending scan operation</a>.
```
Comment (beaufortfrancois) [inline index.html]: I think we should also abort steps below when promise gets rejected right?

File: index.html [hunk 6]
```diff
@@ -3419,6 +3440,9 @@ <h3><dfn>Writing content</dfn></h3>
             browsing context</a>, then reject |p| with and
             {{"InvalidStateError"}} {{DOMException}} and return |p|.
           </li>
+          <li>
+            Attempt to <a>abort a pending scan operation</a>.
```
Comment (beaufortfrancois) [inline index.html]: Same here, we should return |p| with rejected promise if failing.

Comment (zolkis) [reply index.html]: No, we just reject the pending scan, not this one. It's a cleanup operation.
Or do you mean if the abort fails, we should reject the scan?
i.e. should we handle the "ongoing operation" step as a fail to abort and propagate it up?

File: index.html (hunk 5 above)
Comment (zolkis) [reply index.html]: I don't understand.

**Comment by beaufortfrancois** [comment]
To be clear, here are [web platform tests](https://github.com/web-platform-tests/wpt/blob/master/web-nfc/NDEFReader_scan.https.html) for this issue:

```js

nfc_test(async (t, mockNFC) => {
  const ndef = new NDEFReader();
  const promise1 = ndef.scan();
  const promise2 = promise_rejects_dom(t, 'InvalidStateError', ndef.scan());
  await promise1;
  await promise2;
}, "Test that NDEFReader.scan rejects if there is already an ongoing scan.");

nfc_test(async (t, mockNFC) => {
  const ndef = new NDEFReader();
  const controller = new AbortController();
  await ndef.scan({signal : controller.signal});
  controller.abort();
  await ndef.scan();
}, "Test that NDEFReader.scan can be started after the previous scan is aborted.");
```

**Comment by beaufortfrancois** [comment]
@zolkis rightly pointed out that write behaviour is different:

```js
nfc_test(async (t, mockNFC) => {
  const ndef1 = new NDEFReader();
  const ndef2 = new NDEFReader();

  const p1 = ndef1.write(test_text_data, {overwrite: false});
  const p2 = ndef2.write(test_url_data, {overwrite: true});

  await new Promise((resolve, reject) => {
    // Make first push pending
    mockNFC.setPendingPushCompleted(false);
    let err = "";
    p1.then(() => {
      reject("pending push should not be fulfilled");
    }).catch(e => {
      err = e.name;
    });
    p2.then(() => {
      assertNDEFMessagesEqual(test_url_data, mockNFC.pushedMessage());
      assertNDEFWriteOptionsEqual( {overwrite: true}, mockNFC.writeOptions());
      assert_equals(err, "AbortError", "the pending push should be aborted");
      resolve();
    });
  });
}, "NDEFReader.write should replace all previously configured write operations.");
```

**Comment by zolkis** [comment]
I think it makes more sense to replace the previous ops with the current ones, rather than failing the current ones if old ones are ongoing. Saves an explicit abort from the app script. But I am open to be convinced.

**Comment by zolkis** [comment]
What's next? Merge this and update tests?

**Comment by beaufortfrancois** [comment]
Consistency with `write()` behaviour seems better to me. @riju @kenchris WDYT?

**Comment by kenchris** [comment]
+1 consistency

**Comment by zolkis** [comment]
Consistency which way?

If we align to write behavior, this PR is correct. As the title also tells, scan rejects prior scans, i.e. replaces them.

**Comment by beaufortfrancois** [comment]
In example below, how does user know that `ndef1` is not listening to reading events anymore? Or is this what we want?

```js
const ndef1 = new NDEFReader();
await ndef1.scan(); // Promise resolves and reading occurs...
ndef1.onreading => ({message}) => { /* Handle NDEF message */}

// few seconds later
const ndef2 = new NDEFReader();
await ndef2.scan(); // Promise resolves and reading occurs
ndef2.onreading => ({message}) => { /* Handle NDEF message */}

// ndef2 will get reading events.
// How about ndef1?
```

**Comment by zolkis** [comment]
Valid observations. 

In this version of the API, and in this PR, if ndef1.scan() has already resolved and ndef2.scan() is called, the events are NOT coming through ndef1's event since the "abort a pending scan" algorithm steps.

But we don't have a way to notify ndef1 about it having been stopped. Using the onreadingerror event would be a misuse of it. Another pain point in this design. Could be avoided if we had a direct API like resolving the scan() promise with the data read. However, that would not support reading multiple tags (like a design with events does), unless that promise resolved with an array of NDEFMessage's.

(Sorry, this was edited twice).

**Comment by zolkis** [comment]
On a second thought, we could use an `AbortError` reported via `onreadingerror` to signal scan cancellation by a newer scan. At least scripts can react to that (even though it was the same script that made that condition by invoking a newer scan, so the point is kind of moot). Should I include that step in the "abort a pending scan" algo or should we continue with the current way of silently canceling `onreading` event delivery?

**Comment by zolkis** [comment]
According to the discussion in #592, at this point of time people prefer to fail scans if there is an existing scan, but new writes will continue replace pending writes. The reason is that scans report data through an event, which is different from writes.

**Comment by beaufortfrancois** [comment]
I believe the `pending read tuple` should be per reader.
This is not what I'm understanding with latest patch.

See JS code below for my expectations.

 ```js
const ndef1 = new NDEFReader();
await ndef1.scan(); // Promise resolves and reading occurs...
ndef1.onreading => ({message}) => { /* Handle NDEF message */}

// few seconds later
const ndef2 = new NDEFReader();
await ndef2.scan(); // Promise resolves and reading occurs
ndef2.onreading => ({message}) => { /* Handle NDEF message */}

// ndef2 will get reading events.
// ndef1 will also continue to get reading events.
```

```js
const ndef = new NDEFReader();
const p1 = ndef.scan(); // Promise resolves.
const p2 = ndef.scan(); // Promise rejects right away as there's already a pending scan request
```

**Comment by beaufortfrancois** [comment]
@riju can you check my sayings?

**Comment by zolkis** [comment]
> I believe the pending read tuple should be per reader.

No, the spec has stated for ages it was on the settings object, i.e. tied to NfcProxy / NfcImpl, not the readers.
https://w3c.github.io/web-nfc/#nfc-state-associated-with-the-settings-object

The scan sequence diagram also reinforces that.
https://sites.google.com/a/chromium.org/dev/developers/design-documents/web-nfc

**Comment by zolkis** [comment]
> I believe the pending read tuple should be per reader. 

There cannot be a pending read tuple per reader, since then we would store the promise + reader in the reader, which makes no sense.

What we could do in the spec is return to the "activated reader list" internal slot, since the implementation still has that one (meaning the "pending read tuple" internal slot was never implemented).

In the Chromium implementation, `NfcProxy` already has a `DCHECK()` guard to not include the same reader twice. So all policies have to be made at `NfcReader` level. This means the NFC spec has not been implemented according how it was specified (internal slot for pending reader at the settings object). `NfcReader` and `NfcProxy` still allow multiple readers, so it goes to the Android side that way. In `NfcImpl`, the `watch()` function it will return an error if the id already exists, so a repeated `scan()` will fail. Therefore the `has_pending_scan_request_` is not needed, but it's an optimization. In the end, the implementation still has the "activated reader list" and tag content will be dispatched to all readers.

For now, I will try to align the spec to the existing implementation, but this will need to be cleaned up in the future. 
That would satisfy the examples of @beaufortfrancois above.

**Comment by zolkis** [comment]
Yet another change in this PR scope (and name). Now we allow multiple readers, but at each reader, successive scan() invocations will fail.

**Comment by zolkis** [comment]
@riju @kenchris please check.

**Comment by beaufortfrancois** [comment]
@zolkis Please have a look at https://chromium-review.googlesource.com/c/chromium/src/+/2297559/5/third_party/blink/renderer/modules/nfc/ndef_reader.cc and ask help from @riju, the `has_pending_scan_request_` check is only for when the ndef reader didn't settle yet. I believe* this PR doesn't reflect @riju's CL.

**Comment by zolkis** [comment]
@beaufortfrancois 

> has_pending_scan_request_ check is only for when the ndef reader didn't settle yet

Do you refer to this? https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/nfc/ndef_reader.cc;drc=aca54053714d442febdd5f87a26c4055e20bd632;l=187 
(it's another question what is the relation with the similarly named NDEFReader::HasPendingActivity()...)

But then see that the condition is bound to the existence of the reader _after_ it checked was there a pending scan request. Also the design doc states there is only one scan in a reader active at any time.
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/nfc/ndef_reader.cc;drc=aca54053714d442febdd5f87a26c4055e20bd632;l=135 

Could you point out which algorithmic steps are in conflict with the implementation?

**Comment by zolkis** [comment]
We had a call with Arno and Riju, and that clarified things. I discovered a non-related bug in the scan algo, fixed in the commit above.

The `has_pending_scan_request_` flag indeed covers the time between the `scan()` invocation and its promise resolving, i.e. until the `reader` is added to the activated reader objects.
Arno tried removing the flag and resulted in the renderer crashing at a second invocation of `scan()` because it ran into the `DCHECK()` in the beginning of `NfcProxy::StartReading()`. Which IMHO is an implementation bug, since it was supposed to fail the promise on that condition, not to crash the renderer.  The `has_pending_scan_request_` seems to be a _workaround_ for that bug, not an algorithmic necessity.

In the scan algorithm in the spec is IMHO correct: the condition on `reader` being in the activated reader list is sufficient since the reading algorithm is triggered on the same condition. Single point of exclusion is desired, so the algorithm is correct. We don't need to factor out a separate internal slot to reflect the current implementation behavior that uses `has_pending_scan_request_`.

The existing WPT should be working as defined with the current specification of the scan algorithm.

File: index.html [hunk 7]
```diff
@@ -3477,7 +3465,12 @@ <h3><dfn>Writing content</dfn></h3>
                 and abort these steps.
               </li>
               <li>
-                Set the <a>pending read tuple</a> to (|p|, |reader|).
+                If |reader| is already in the <a>activated reader objects</a>,
+                then reject |p| with a {{"InvalidStateError"}} {{DOMException}}
```
Comment (beaufortfrancois) [inline index.html]:
```suggestion
                then reject |p| with an {{"InvalidStateError"}} {{DOMException}}
```

**Review by beaufortfrancois — CHANGES_REQUESTED** [review]

**Comment by beaufortfrancois** [comment]
> Arno tried removing the flag and resulted in the renderer crashing at a second invocation of `scan()` because it ran into the `DCHECK()` in the beginning of `NfcProxy::StartReading()`. Which IMHO is an implementation bug, since it was supposed to fail the promise on that condition, not to crash the renderer. The `has_pending_scan_request_` seems to be a _workaround_ for that bug, not an algorithmic necessity.

Crashing will only happen on Debug builds (hence *D*CHECK) and is only for chromium developers to debug non expected cases.

> The existing WPT should be working as defined with the current specification of the scan algorithm.

I believe we should update one web platform test to make sure it's fully working:

```diff
 nfc_test(async (t, mockNFC) => {
   const ndef = new NDEFReader();
   const controller = new AbortController();
   await ndef.scan({signal : controller.signal});
+  await promise_rejects_dom(t, 'InvalidStateError', ndef.scan());
   controller.abort();
   await ndef.scan();
 }, "Test that NDEFReader.scan can be started after the previous scan is aborted.");
```

**Comment by zolkis** [comment]
Thanks for the updates.

> "Test that NDEFReader.scan can be started after the previous scan is aborted."

Description needs update as well? Isn't that testing 3 scans? Tests if the second scan() fails since the first is started, and the third can be started once the first is aborted? Maybe split in 2 test cases?

**Comment by beaufortfrancois** [comment]
> Description needs update as well? Isn't that testing 3 scans? Tests if the second scan() fails since the first is started, and the third can be started once the first is aborted? Maybe split in 2 test cases?

I've ended up updating another test. See https://github.com/web-platform-tests/wpt/pull/27014

**Review by beaufortfrancois — APPROVED** [review]
Thank you @zolkis!

**Pull request merged**